Add non-generic Create method to DispatchProxy

heathbm opened the pull request:
Implementation of #67444

heathbm committed:
Add non-generic Create method to DispatchProxy 

WIP, this commit is to clarify the approach

heathbm commented:
Can the attribute constraint `DynamicallyAccessedMembers` be applied to method parameters of type `Type`.  I would like to contribute if I can. However, I am slightly unsure of the approach to take with regards to these constraints, as `DispatchProxyGenerator.CreateProxyInstance` requires them.

teo-tsirpanis commented:
@heathbm you need to also update https://github.com/dotnet/runtime/blob/main/src/libraries/System.Reflection.DispatchProxy/ref/System.Reflection.DispatchProxy.cs.

teo-tsirpanis commented:
And I think yes the attribute can be applied to `Type` members.

jkotas commented on src/libraries/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxy.cs:
```diff
@@ -39,5 +39,16 @@ protected DispatchProxy()
+        {
+            if (!proxyType.IsAssignableTo(typeof(DispatchProxy)))
+            {
+                throw new ArgumentException("proxyType must be assignable to DispatchProxy");
```
This message should be defined in https://github.com/dotnet/runtime/blob/main/src/libraries/System.Reflection.DispatchProxy/src/Resources/Strings.resx

heathbm replied:
Updated. The new exception message looks like so:  
**System.ArgumentException**: ` The provided proxy type '<YOUR_TYPE>' is not assignable to the type 'DispatchProxy'.`

heathbm committed:
DispatchProxy.Create: Add exception message, add null checks, update ref

heathbm committed:
DispatchProxy.Create: Add tests

Add equivalent tests for all existing generic DispatchProxy.Create tests.
Add new tests for the null checks.

heathbm commented:
> @heathbm you need to also update https://github.com/dotnet/runtime/blob/main/src/libraries/System.Reflection.DispatchProxy/ref/System.Reflection.DispatchProxy.cs.

Updated.

heathbm commented:
> And I think yes the attribute can be applied to `Type` parameters.

@teo-tsirpanis Thank you for confirming.

MichalStrehovsky commented on src/libraries/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxy.cs:
```diff
@@ -39,5 +39,35 @@ protected DispatchProxy()
+                throw new ArgumentNullException(nameof(proxyType));
+            }
+
+            if (!proxyType.IsAssignableTo(typeof(DispatchProxy)))
+            {
+                throw new ArgumentException(SR.Format(SR.ProxyType_Must_Be_Assignable_To_DispatchProxy, proxyType.Name));
+            }
```
This will need more validation - many kinds of bad Type instances can be passed to this overload. For example:

```csharp
// Passes all the validations so far (prints true)
Console.WriteLine(typeof(BadProxy1<>).IsAssignableTo(typeof(DispatchProxy)));
// Passes all the validations so far (prints true)
Console.WriteLine(typeof(BadProxy2<>).GetGenericArguments()[0].IsAssignableTo(typeof(DispatchProxy)));

abstract class BadProxy1<T> : DispatchProxy { }
class BadProxy2<T> where T : DispatchProxy { }
```

We also need to define how this should react if someone passes a Type instance that is not even owned by the runtime. E.g. if it's obtained from a MetadataLoadContext.

buyaa-n replied:
> This will need more validation - many kinds of bad Type instances can be passed to this overload.

Added test scenarios you mentioned, could not find other bad cases, it seems existing tests has quite good coverage.
https://github.com/dotnet/runtime/blob/a4bc7fab9cf9628d8ab7cd1b12e6ab2516ce9046/src/libraries/System.Reflection.DispatchProxy/tests/DispatchProxyTests.cs#L136-L177
The problem here: the argument names passed to ArgumentException is same as generic overload type parameter name, I am thinking to pass the argument names where the exceptions thrown



> We also need to define how this should react if someone passes a Type instance that is not even owned by the runtime. E.g. if it's obtained from a MetadataLoadContext.

I think we should throw an ArgumentException in that case, not sure how to check if a type is loaded by MLC, added a test, it fails on `if (!proxyType.IsAssignableTo(typeof(DispatchProxy)))` condition and throws

steveharter replied:
> We also need to define how this should react if someone passes a Type instance that is not even owned by the runtime. E.g. if it's obtained from a MetadataLoadContext.

If a type is owned by MetadataLoadContext then its base class (i.e. `DispatchProxy`) will also be a separate type from the runtime `DispatchProxy` type and thus doesn't need to be specially checked since `IsAssignableTo` will return `false`.

steveharter replied:
If we need it, I suppose a check like `typeof(object).GetType() == proxyType.GetType()` would differentiate between RuntimeType and others (RoType).

jkotas replied:
It is not safe to assume that all runtime type objects are of the same type.

buyaa-n replied:
`proxyType.Assembly.ReflectionOnly` works for existing tests, but not sure if `Assembly.ReflectionOnly` would always work as ReflectionOnly loading is deprecated

steveharter replied:
To recap where I think we are here:

There are 2 Types to validate: `proxyType` and `interfaceType`.

For `proxyType` should just be able to do `proxyType.IsAssignableTo(typeof(DispatchProxy))` which should work fine with MLC and other cases where the base Type system is custom.

For `interfaceType`, if we don't validate it, we will eventually get an exception because there is no `Type.TypeHandle`. So if we still want to validate this, we should use an established pattern assuming there is one and that we've validated types before in similar cases. 

@jkotas do you know what the pattern is to validate that a Type is a valid runtime type? Thoughts:
- Use `type.TypeHandle` and then validate a non-default value. MLC throws InvalidOperation so we probably just let it throw without try\catch.
- Use `type is RuntimeType` or `type.TypeHandle is RuntimeTypeHandle`. However, since those types are internal only works for internal\core code unless we find\add a wrapper. 
- Use `typeof(object).GetType() == type.GetType()`. May not work in all cases?
- Use `proxyType.Assembly.ReflectionOnly` (probably not since deprecated)

buyaa-n replied:
Thanks for the recap, that is exactly what I need

jkotas replied:
I do not think we need to bother checking that the passed in type is RuntimeType. We do not do it anywhere else in the libraries and just let it throw exception from within the implementation if it calls unsupported method. For example, try the following with System.Text.Json:
```csharp
using System.Reflection;
using System.Reflection.Emit;
using System.Text.Json;

AssemblyBuilder ab = AssemblyBuilder.DefineDynamicAssembly(new AssemblyName("MyAssembly"), AssemblyBuilderAccess.Run);
ModuleBuilder mb = ab.DefineDynamicModule("MyModule");
TypeBuilder tb = mb.DefineType("MyType", TypeAttributes.Public);

JsonSerializer.Deserialize("{ }", tb);
```
It is going to throw at:
```
Unhandled exception. System.NotSupportedException: The invoked member is not supported before the type is created.
   at System.Reflection.Emit.TypeBuilder.GetCustomAttributes(Type attributeType, Boolean inherit)
   at System.Text.Json.Reflection.ReflectionExtensions.GetUniqueCustomAttribute[TAttribute](MemberInfo memberInfo, Boolean inherit)
   at System.Text.Json.Serialization.Metadata.DefaultJsonTypeInfoResolver.GetConverterForType(Type typeToConvert, JsonSerializerOptions options, Boolean resolveJsonConverterAttribute)
   at System.Text.Json.Serialization.Metadata.DefaultJsonTypeInfoResolver.CreateJsonTypeInfo(Type type, JsonSerializerOptions options)
   at System.Text.Json.Serialization.Metadata.DefaultJsonTypeInfoResolver.GetTypeInfo(Type type, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializerOptions.GetTypeInfoNoCaching(Type type)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at System.Text.Json.JsonSerializerOptions.GetTypeInfoInternal(Type type, Boolean ensureConfigured, Boolean resolveIfMutable)
   at System.Text.Json.JsonSerializer.GetTypeInfo(JsonSerializerOptions options, Type inputType)
   at System.Text.Json.JsonSerializer.Deserialize(String json, Type returnType, JsonSerializerOptions options)
   at Program.<Main>$(String[] args) in C:\repro\Program.cs:line 9
```
The reflection object model has been (at least partially) designed to allow custom types pretending to be runtime types so allowing custom types to be passed in is not necessarily wrong.

jkotas replied:
Some comments on the proposed options:

> Use type.TypeHandle and then validate a non-default value. MLC throws InvalidOperation so we probably just let it throw without try\catch.
> Use proxyType.Assembly.ReflectionOnly (probably not since deprecated)

The custom types can override these methods and return whatever from them, so this would be assuming that custom type is implemented correctly. You can as well assume that the custom type is implemented correctly to make everything else work.

> Use type is RuntimeType or type.TypeHandle is RuntimeTypeHandle. However, since those types are internal only works for internal\core code unless we find\add a wrapper.

Right, you would need to have access to runtime internals.

> Use typeof(object).GetType() == type.GetType(). May not work in all cases?

This depends on the runtime implementain details. It happens to work for the current CoreCLR reflection implementation, but it won't work for current native AOT reflection implementation.

buyaa-n replied:
> I do not think we need to bother checking that the passed in type is RuntimeType. We do not do it anywhere else in the libraries and just let it throw exception from within the implementation if it calls unsupported method.

Sounds good to me thanks, will remove new test with MLC

stephentoub commented on src/libraries/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxy.cs:
```diff
@@ -39,5 +39,35 @@ protected DispatchProxy()
+        [RequiresDynamicCode("Creating a proxy instance requires generating code at runtime")]
+        public static object Create([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] Type interfaceType, [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)] Type proxyType)
+        {
+            if (interfaceType == null)
+            {
+                throw new ArgumentNullException(nameof(interfaceType));
+            }
+
+            if (proxyType == null)
+            {
+                throw new ArgumentNullException(nameof(proxyType));
+            }
+
+            if (!proxyType.IsAssignableTo(typeof(DispatchProxy)))
+            {
+                throw new ArgumentException(SR.Format(SR.ProxyType_Must_Be_Assignable_To_DispatchProxy, proxyType.Name));
+            }
```
```suggestion
            ArgumentNullException.ThrowIfNull(interfaceType);
            ArgumentNullException.ThrowIfNull(proxyType);
            if (!proxyType.IsAssignableTo(typeof(DispatchProxy)))
            {
                throw new ArgumentException(SR.Format(SR.ProxyType_Must_Be_Assignable_To_DispatchProxy, proxyType.Name), nameof(proxyType));
            }
```

joperezr commented:
[Triage]: @heathbm looks like there are still some un-addressed comments here, are you still planning on addressing those?

heathbm commented:
> [Triage]: @heathbm looks like there are still some un-addressed comments here, are you still planning on addressing those?

@joperezr I still intend to address these. However, I have not had time recently. I would be happy to collaborate with others if someone wants to contribute here. If not, I will get to these eventually.

buyaa-n committed:
Apply feedbacks, add more test

buyaa-n committed:
Rever unwanted change

buyaa-n committed:
Pass correct parameter name into ArgumentException, add more test case

jkotas commented on src/libraries/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxy.cs:
```diff
@@ -11,6 +11,7 @@ namespace System.Reflection
     /// </summary>
     public abstract class DispatchProxy
     {
+        private static readonly Type s_dispatchProxyType = typeof(DispatchProxy);
```
Caching `typeof` like this is unnecessary. It is actually going to be de-optimization with recent JIT changes.

buyaa-n committed:
Revert caching DispatchProxy type

jkotas commented on src/libraries/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxyGenerator.cs:
```diff
@@ -168,35 +170,34 @@ public ProxyAssembly(AssemblyLoadContext alc)
 
                 // The interface type must be an interface, not a class
-                if (!interfaceType.IsInterface)
+                if (!interfaceType.IsInterface || interfaceType.Assembly.ReflectionOnly)
```
```suggestion
                if (!interfaceType.IsInterface)
```

buyaa-n committed:
Remove MLC related test and related code
